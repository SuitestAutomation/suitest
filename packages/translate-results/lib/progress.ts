import t from './texts';
import {Translation} from './types';
import {StartupError, translateStartupError} from './startup';

/**
 * @description Get humanized progress status explanation result
 */
export function translateProgress(message: ProgressMessage): Translation {
	if ('code' in message) {
		return translateStartupError(message.code);
	}

	switch (message.status) {
		case 'openingApp': return {title: t['progress.status.openingApp']()};
		case 'closingApp': return {title: t['progress.status.closingApp']()};
		case 'bootingDevice': return {title: t['progress.status.bootingDevice']()};
		case 'needManual': return {title: t['progress.status.needManual']()};
		case 'recoveringID': return {title: t['progress.status.recoveringID']()};
		case 'waitingForConnectionFromBootstrap': return {title: t['progress.status.waitingForConnectionFromBootstrap']()};
		case 'waitingForConnectionFromIL': return {title: t['progress.status.waitingForConnectionFromIL']()};
		case 'unistallingApp': return {title: t['progress.status.unistallingApp']()};
		case 'uploadingAndInstallingApp': return {title: t['progress.status.uploadingAndInstallingApp']()};
		case 'nothing': break;
		// actionFailed should comes with code, so it will proceeded by translateStartupError above
		case 'actionFailed': break;
		default:
			const _status: never = message.status;
			console.warn(_status, 'progression status not handled');
	}

	return {title: ''};
}

export const PROGRESS_STATUS = Object.freeze({
	actionFailed: 'actionFailed',
	openingApp: 'openingApp',
	closingApp: 'closingApp',
	bootingDevice: 'bootingDevice',
	needManual: 'needManual',
	nothing: 'nothing',
	recoveringID: 'recoveringID',
	waitingForConnectionFromBootstrap: 'waitingForConnectionFromBootstrap',
	waitingForConnectionFromIL: 'waitingForConnectionFromIL',
	unistallingApp: 'unistallingApp',
	uploadingAndInstallingApp: 'uploadingAndInstallingApp',
} as const);

export type ProgressStatus = keyof typeof PROGRESS_STATUS;

type ProgressMessage = {
	status?: typeof PROGRESS_STATUS.actionFailed
	code: StartupError,
} | {
	status: ProgressStatus,
};
