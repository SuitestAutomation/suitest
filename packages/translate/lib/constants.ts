export const NOT_STARTED_REASON = Object.freeze({
	BLASTER_ERROR: 'blasterError',
	BOOTING: 'runningBootSequence',
	CANDYBOX_OFFLINE: 'candyBoxOffline',
	DEVICE_DELETED: 'deviceDeleted',
	DEVICE_DISABLED: 'deviceDisabled',
	DEVICE_IN_USE: 'deviceInUse',
	EXCEEDED_PLAN_LIMITS: 'planLimitExceeded',
	INTERNAL_ERROR: 'internalError',
	NO_ACTIVE_PLAN: 'noActivePlan',
	NO_AUTOMATED_MINUTES: 'noAvailableAutomatedMinutes',
	PLATFORM_ERROR_ANDROID: 'androidPlatformError',
	PLATFORM_ERROR_BOOTSTRAP: 'bootstrappedPlatformError',
	PLATFORM_ERROR_LG_WEBOS: 'lgWebosPlatformError',
	PLATFORM_ERROR_UNDEFINED: 'notDefinedPlatform',
	PLATFORM_ERROR_PLAYSTATION: 'playstationPlatformError',
	PLATFORM_ERROR_ROKU: 'rokuPlatformError',
	PLATFORM_ERROR_XBOX: 'xboxPlatformError',
	PLATFORM_ERROR_APPLE_TV: 'applePlatformError',
	PLATFORM_ERROR_TIZEN: 'tizenPlatformError',
	SUITEST_DRIVE_OFFLINE: 'suitestDriveOffline',
	SUITEST_DRIVE_SERVICE_OFFLINE: 'suitestDriveServiceOffline',
	TEST_QUEUED: 'testQueued',
	PACKAGE_CORRUPTED: 'packageCorrupted',
	CONFIGURATOR_ERROR: 'configuratorError',
	APPLE_NETWORK_LOGS_ERROR: 'appleNetworkLogsError',
	DEVICE_OFFLINE: 'deviceOffline',
	SUITEST_DRIVE_UPDATE: 'suitestDriveUpdate',
} as const);

export type NotStartedReason = typeof NOT_STARTED_REASON[keyof typeof NOT_STARTED_REASON];

export const PROGRESS_STATUS = Object.freeze({
	ACTION_FAILED: 'actionFailed', // says that previous operation failed
	APP_UNINSTALL: 'unistallingApp',
	APP_UPLOAD_INSTALL: 'uploadingAndInstallingApp',
	BOOTING_DEVICE: 'bootingDevice',
	CLOSING_APP: 'closingApp',
	DEVICE_IDENTIFICATION: 'recoveringID',
	NOTHING: 'nothing', // says that previous operation is finished (failed or succeed)
	OPENING_APP: 'openingApp',
	WAITING_FOR_BOOTSTRAP: 'waitingForConnectionFromBootstrap',
	WAITING_FOR_IL: 'waitingForConnectionFromIL',
	WAIT_FOR_MANUAL_ACTION: 'needManual',
} as const);

export type ProgressStatus = typeof PROGRESS_STATUS[keyof typeof PROGRESS_STATUS];
