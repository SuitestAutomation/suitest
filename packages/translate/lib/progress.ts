import t from './texts';
import {Translation} from './types';
import {PROGRESS_STATUS, ProgressStatus, NotStartedReason} from './constants';
import {translateNotStartedReason} from './notStartedReason';

/**
 * @description Get humanized progress status explanation result
 */
export function translateProgress(message: ProgressMessage): Translation | undefined {
	if ('code' in message && message.code !== undefined) {
		return translateNotStartedReason(message.code);
	}

	switch (message.status) {
		case PROGRESS_STATUS.OPENING_APP: return {title: t['progress.status.openingApp']()};
		case PROGRESS_STATUS.CLOSING_APP: return {title: t['progress.status.closingApp']()};
		case PROGRESS_STATUS.BOOTING_DEVICE: return {title: t['progress.status.bootingDevice']()};
		case PROGRESS_STATUS.WAIT_FOR_MANUAL_ACTION: return {title: t['progress.status.needManual']()};
		case PROGRESS_STATUS.DEVICE_IDENTIFICATION: return {title: t['progress.status.recoveringID']()};
		case PROGRESS_STATUS.WAITING_FOR_BOOTSTRAP: return {title: t['progress.status.waitingForConnectionFromBootstrap']()};
		case PROGRESS_STATUS.WAITING_FOR_IL: return {title: t['progress.status.waitingForConnectionFromIL']()};
		case PROGRESS_STATUS.APP_UNINSTALL: return {title: t['progress.status.unistallingApp']()};
		case PROGRESS_STATUS.APP_UPLOAD_INSTALL: return {title: t['progress.status.uploadingAndInstallingApp']()};
		case PROGRESS_STATUS.NOTHING: return {title: ''};
		// actionFailed should come with code, so it will be proceeded by translateStartupError above
		case 'actionFailed': break;
		default:
			const _status: never = message.status;
			console.warn(_status, 'progress status not handled');
	}

	return undefined;
}

type ProgressMessage = {
	status?: typeof PROGRESS_STATUS.ACTION_FAILED,
	code: NotStartedReason,
} | {
	status: ProgressStatus,
	code?: undefined,
};
