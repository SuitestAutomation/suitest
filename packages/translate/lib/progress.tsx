import {Translation} from './types';
import {PROGRESS_STATUS, ProgressStatus, NotStartedReason} from './constants';
import {translateNotStartedReason} from './notStartedReason';
import {jsx} from '@suitest/smst';

/**
 * @description Get humanized progress status explanation result
 * @throws {Error} Throws an Error if unknown status is provided or code is required and missing
 */
export function translateProgress(message: ProgressMessage): Translation {
	if ('code' in message && message.code !== undefined) {
		return translateNotStartedReason(message.code);
	}

	if (message.status === undefined) {
		throw new Error('A status is expected.');
	}

	switch (message.status) {
		case PROGRESS_STATUS.OPENING_APP: return {
			title: <text>Trying to open app…</text>,
		};
		case PROGRESS_STATUS.CLOSING_APP: return {
			title: <text>Trying to close app…</text>,
		};
		case PROGRESS_STATUS.BOOTING_DEVICE: return {
			title: <text>Running the boot sequence defined for the device…</text>,
		};
		case PROGRESS_STATUS.WAIT_FOR_MANUAL_ACTION: return {
			title: <text>Paused. For this platform, install and open the application manually.</text>,
		};
		case PROGRESS_STATUS.DEVICE_IDENTIFICATION: return {
			title: <text>Trying to recover Suitest device ID…</text>,
		};
		case PROGRESS_STATUS.SELECTING_SUITESTIFY_INSTANCE: return {
			title: <text>Selecting nearest suitestify instance…</text>,
		};
		case PROGRESS_STATUS.WAITING_FOR_BOOTSTRAP: return {
			title: <text>Waiting for connection from the Suitest app on device…</text>,
		};
		case PROGRESS_STATUS.WAITING_FOR_IL: return {
			title: <text>Waiting for connection from the instrumentation library…</text>,
		};
		case PROGRESS_STATUS.APP_UNINSTALL: return {
			title: <text>Uninstalling app…</text>,
		};
		case PROGRESS_STATUS.APP_UPLOAD_INSTALL: return {
			title: <text>Uploading and installing app…</text>,
		};
		case PROGRESS_STATUS.NOTHING:
		case PROGRESS_STATUS.ACTION_FAILED: return {
			title: <text>{''}</text>,
		};
		default:
			const _status: never = message.status;
			throw new Error(`Unknown status code received: ${_status}`);
	}
}

type ProgressMessage = {
	code?: NotStartedReason,
	status?: ProgressStatus,
};
