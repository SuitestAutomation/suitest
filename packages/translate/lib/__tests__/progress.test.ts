import {PROGRESS_STATUS} from '../constants';
import {translateProgress} from '../progress';

describe('Interactive progress explanation.', () => {
	// No need for deep testing because translateProgress with code will return the same result as translateStartupError
	it('Should return translation if code is valid', () => {
		const res = {
			title: 'Cannot continue: Your subscription has expired',
			description: 'Your subscription has expired, to continue using Suitest please [renew your subscription](https://the.suite.st/preferences/billing)',
		};
		expect(translateProgress({code: 'noActivePlan'})).toStrictEqual(res);
		expect(translateProgress({code: 'noActivePlan', status: 'actionFailed'})).toStrictEqual(res);
	});

	for (const status of Object.values(PROGRESS_STATUS)) {
		if (status === 'actionFailed') {
			continue;
		}

		it(`Should translate "${status}" status`, () => {
			expect(translateProgress({status})).toMatchSnapshot();
		});
	}

	it('Should throw an error if actionFailed message does not deliver code', () => {
		expect(() => translateProgress({status: 'actionFailed'}))
			.toThrowError('A code is expected when receiving "actionFailed" progress status');
	});

	it('Should handle unknown statuses/codes', () => {
		expect(() => translateProgress({status: 'unknownStatus'} as any))
			.toThrowError('Unknown status code received: unknownStatus');

		expect(() => translateProgress({status: 'unknownStatus', code: 'unknownCode'} as any))
			.toThrowError('Unknown not started reason received: unknownCode');
		expect(() => translateProgress({code: 'unknownCode'} as any))
			.toThrowError('Unknown not started reason received: unknownCode');

		// status valid and code undefined
		expect(translateProgress({status: 'openingApp', code: undefined})).toStrictEqual({
			title: 'Trying to open appâ€¦',
		});
		// code is valid and status undefined
		expect(translateProgress({status: undefined, code: 'candyBoxOffline'})).toStrictEqual({
			description: 'Check that the cable plugged into the CandyBox delivers Internet connection or reboot the CandyBox and allow about 5 minutes for it to initialize',
			title: 'Cannot continue: CandyBox controlling this device is offline',
		});
	});
});
