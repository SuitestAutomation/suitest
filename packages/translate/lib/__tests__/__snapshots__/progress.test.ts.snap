// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Interactive progress explanation. Should translate "bootingDevice" status 1`] = `
Object {
  "title": "Running the boot sequence defined for the device…",
}
`;

exports[`Interactive progress explanation. Should translate "closingApp" status 1`] = `
Object {
  "title": "Trying to close app…",
}
`;

exports[`Interactive progress explanation. Should translate "needManual" status 1`] = `
Object {
  "title": "Paused. For this platform, install and open the application manually.",
}
`;

exports[`Interactive progress explanation. Should translate "nothing" status 1`] = `
Object {
  "title": "",
}
`;

exports[`Interactive progress explanation. Should translate "openingApp" status 1`] = `
Object {
  "title": "Trying to open app…",
}
`;

exports[`Interactive progress explanation. Should translate "recoveringID" status 1`] = `
Object {
  "title": "Trying to recover Suitest device ID…",
}
`;

exports[`Interactive progress explanation. Should translate "unistallingApp" status 1`] = `
Object {
  "title": "Uninstalling app…",
}
`;

exports[`Interactive progress explanation. Should translate "uploadingAndInstallingApp" status 1`] = `
Object {
  "title": "Uploading and installing app…",
}
`;

exports[`Interactive progress explanation. Should translate "waitingForConnectionFromBootstrap" status 1`] = `
Object {
  "title": "Waiting for connection from the Suitest app on device…",
}
`;

exports[`Interactive progress explanation. Should translate "waitingForConnectionFromIL" status 1`] = `
Object {
  "title": "Waiting for connection from the instrumentation library…",
}
`;
